### PR TITLE
fix: fixes for downloading

### LINKED ISSUE #507: [iOS] Changing the number of files available for download
When canceling a download and deleting already downloaded content, sometimes several previously available files for download are lost.
When canceling a download and deleting already downloaded content, sometimes several previously available files for download are lost, they are no longer taken into account, and cannot be deleted or downloaded again. How to repeat: start bulk download ->waiting while 1/4 of the part to download -> stop the download and delete the downloaded content (using switch) -> of the 99 files initially available, only 98 remain available

Image

### PR DESCRIPTION
**Pull request by rnr**
This PR fixes next issues:
https://github.com/openedx/openedx-app-ios/issues/485
https://github.com/openedx/openedx-app-ios/issues/458

To fix the remaining tmp files, DownloadRequest.Destination was added with the setting .removePreviousFile.
The way we cancel downloads at the chapter level has also been changed: since we have no individual control at this level, we must cancel and remove all files in the chapter to prevent a situation where we cancel a download but the files remain.

### CODE REVIEW

**Comment by saeedbashir** [comment]
I'm not sure whether this is related to this, but downloaded videos are not deleting.
https://github.com/user-attachments/assets/23205715-2c4c-4952-b8f0-9a265153e2a4

**Comment by rnr** [comment]
> I'm not sure whether this is related to this, but downloaded videos are not deleting. https://github.com/user-attachments/assets/23205715-2c4c-4952-b8f0-9a265153e2a4

@saeedbashir Do you see it in this branch only? Could you please provide some details to reproduce this. Thank you

**Comment by saeedbashir** [comment]
> @saeedbashir Do you see it in this branch only? Could you please provide some details to reproduce this. Thank you

I was playing around with the downloads and it happened. Now, I'm not able to replicate that issue but I am facing another issue, the downloads continue even after canceling the download. 

https://github.com/user-attachments/assets/c651dd54-dd57-48fe-b371-dcc3bfad30b7

File: Core/Core/Network/DownloadManager.swift [hunk 1]
```diff
@@ -268,7 +268,8 @@ public class DownloadManager: DownloadManagerProtocol {
     public func deleteFile(blocks: [CourseBlock]) async {
         for block in blocks {
             do {
-                if let fileURL = fileUrl(for: block.id) {
+                if let fileURL = fileUrl(for: block.id),
```
Comment (saeedbashir) [inline Core/Core/Network/DownloadManager.swift]: Getting `Trailing Whitespace Violation:` warning here.

Comment (rnr) [reply Core/Core/Network/DownloadManager.swift]: fixed

**Comment by rnr** [comment]
> > @saeedbashir Do you see it in this branch only? Could you please provide some details to reproduce this. Thank you
> 
> I was playing around with the downloads and it happened. Now, I'm not able to replicate that issue but I am facing another issue, the downloads continue even after canceling the download.
> 
>  Screen.Recording.2024-07-18.at.8.17.16.AM.mov

@saeedbashir Yes we are facing this bug too - it's not just cancelling issue it's how downloads start. @forgotvas is looking on this. We will create ticket/issue and perhaps separate branch/PR will be better for this. Thank you

**Comment by saeedbashir** [comment]
> > > @saeedbashir Do you see it in this branch only? Could you please provide some details to reproduce this. Thank you
> > 
> > 
> > I was playing around with the downloads and it happened. Now, I'm not able to replicate that issue but I am facing another issue, the downloads continue even after canceling the download.
> > Screen.Recording.2024-07-18.at.8.17.16.AM.mov
> 
> @saeedbashir Yes we are facing this bug too - it's not just cancelling issue it's how downloads start. @forgotvas is looking on this. We will create ticket/issue and perhaps separate branch/PR will be better for this. Thank you

If that is the case, we can merge this PR. Please create a issue for this so it wont lost and we have to fix it before MVP.

**Review by saeedbashir — APPROVED** [review]
Approving this as per the discussion 👍

**Review by forgotvas — APPROVED** [review]

**Comment by rnr** [comment]
Sergey made these tickets:
https://github.com/openedx/openedx-app-ios/issues/508
https://github.com/openedx/openedx-app-ios/issues/506
merging this

**Pull request merged**